### ISSUE TITLE
Unicorn Plugin

### ISSUE THREAD

**Issue by Manouchehri** [angr/simuvex#29]
As previously mentioned in https://github.com/angr/cle/issues/22#issuecomment-221132785, a [Unicorn Engine](http://www.unicorn-engine.org/) plugin is in progress. Thought it would make sense to open up a GitHub issue in the proper repo so other people are aware.

@isra17 recently posted a neat benchmark showing angr vs. Unicorn. https://github.com/isra17/emu_test

**Comment by rhelmot** [comment]
(As a clerical note, it looks likely at this point that we will not release
the unicorn support until after the CGC in August)

**Comment by zardus** [comment]
Yep, we're definitely working on it! It won't quite match @isra17's unicorn benchmarks (because we need to hook memory accesses to sync the state back into angr), but it'll be a *lot* faster than emulating VEX with python. Some preliminary benchmarks show runtime increases of *600x* for certain workloads.

**Comment by Manouchehri** [comment]
Exciting stuff! Looking forward to the release after you win CGC. =)

**Comment by ltfish** [comment]
I guess we will release it even if we don't win CGC :D

**Comment by aquynh** [comment]
nice to hear, and looking forward to your release this Unicorn support :-)

**Comment by Manouchehri** [comment]
For anyone following this, Unicorn support has been completed but not pushed to the repo yet.

**Comment by aquynh** [comment]
cool, cant wait! but exactly how Unicorn is used there?

**Comment by Manouchehri** [comment]
As per @zardus's previous comment, it's used instead of emulating VEX with Python.

**Comment by salls** [comment]
The main use cases of unicorn right now are the following:
1. The data currently being operated on is concrete. Symbolic emulation isn't necessary so we jump into unicorn and let unicorn/qemu execute until symbolic data would be touched or introduced.
2. There is something we haven't implemented (weird vex instruction, syscall etc). In these cases I believe it is better to run the part angr can't handle with unicorn. We lose the symbolic relationship, but at least we can continue execution in a sane state

**Comment by Manouchehri** [comment]
Looks like it's been pushed!

**Comment by rhelmot** [comment]
It's been pushed for the last two weeks! The big thing, though, is that our fork of Unicorn is now pushed, to https://github.com/angr/unicorn. 

The changes we've made are all pending pull requests to upstream. They're mostly bug fixes and minor features, the biggest thing is that it overhauls the python setup process to do things like actually work correctly without manually copying files around, and also packaging the library headers so simuvex can build against it.

Basically, you should just be able to `python setup.py install` in the bindings/python directory and it should Just Work. (You can also run the setup.py in the repo root but that's just a compatibility shim until it's distributed on pip)
